File: lootcard.py
"""What a drop is worth showing on the stream, and for how long.

infra#3334, follow-up. When something drops, a viewer sees a small roll frame
with a truncated name and a chat line that scrolls away. What dropped, whether
it is any good, and who got it are the whole moment, and all three are
invisible unless somebody hovers the icon. Nobody is at the keyboard to hover.

THE STATS ARE THE GAME'S, NOT OURS. The addon calls
`GameTooltip:SetHyperlink(link)` on the link that came through chat and the
client renders its own tooltip, so a suffix, an enchant or a random stat roll
is right because the client drew it. Nothing here reconstructs an item, and
nothing here ever should: a display that composes its own stats is a display
that starts lying the first time an item is not the plain case.

WHAT THIS MODULE DECIDES, which is everything a test can reach:

  which drops are worth a card    the quality bar, below
  what the headline says          "Og won", "Grug looted"
  how long a card stays           DWELL_SECONDS
  how many can be up at once      MAX_CARDS, and how many keep their stats
  which chat lines are loot       MESSAGE_FORMS, and the ORDER to try them in
  when two lines are one drop     DEDUPE_SECONDS

The addon reads the format strings out of the client's own globals and turns
them into patterns. That part cannot move here - `LOOT_ITEM` only exists inside
a running client, and hardcoding "%s receives loot: %s." in English is exactly
what those globals exist to prevent. What CAN live here is the table: which
globals, in what order, and which argument of each is the winner and which is
the item. tests/test_lootcard.py holds the Lua to that table as source text.

THE ORDER IS NOT ALPHABETICAL AND IS NOT DECORATION. "%s won: %s" matches the
"|cff818181(Need - 76)|r" line too, with the suffix swallowed into the item
capture, and it matches "You won: ..." with a winner called "You". So the
specific forms are tried before the general ones and the self forms before the
third-person ones. Reorder this list and the wrong capture wins silently.

A DROP IS ONE MOMENT AND THE SERVER SENDS IT TWICE. Group loot writes the roll
result ("Og won: [x]") and then the loot itself ("Og receives loot: [x]."), so
the same robe would open two cards seconds apart. They collapse, and the WON
wording beats LOOTED because a contested item that somebody actually won is the
more interesting sentence.

ROLLS IN PROGRESS ARE NOT SHOWN. `START_LOOT_ROLL` already puts a roll frame on
screen with its own timer; a card for the roll and another for its result would
double the screen time of every drop to say the same thing twice. The moment
worth showing is the resolution.

PURE MODULE: facts in, decisions out. No client, no clock of its own unless one
is not handed to it. Nothing in the images imports it yet - it exists to be
the one place these rules are written down and the one place they are tested
- but it ships with the rest anyway, because the manifest is all or nothing:
tests/test_ship_manifest.py requires every top-level module to be named in
the Dockerfile, after a rerere replay once dropped panel.py from it and gave
the map server an import crash at pod start, long after CI was green.
"""
from __future__ import annotations

# The bar. UNCOMMON, which is 2, and the number is the operator's complaint
# rather than taste: a stream that announces every scrap of wool and every
# Moss Agate is noise with a robe hidden in it. Greens and better are the
# things a level-25 party actually stops for.
#
# AN ITEM WHOSE QUALITY CANNOT BE READ IS NOT SHOWN, and that direction is
# chosen deliberately. Quality comes out of the link's own colour code, so a
# link we cannot classify means the reader is broken rather than that one
# unusual item arrived - and a broken reader that shows everything would put
# every grey on a permanent stream, while a broken reader that shows nothing
# costs one missed card and is obvious.
MIN_QUALITY = 2

# A couple of minutes, taken literally. Long enough that a viewer who looked
# away at the pull still sees what dropped; short enough that a card is gone
# before the next pull is over.
DWELL_SECONDS = 120

# How many cards may be up, and how many of those keep their tooltip.
#
# BOTH NUMBERS COME FROM THE SCREEN. A tooltip for a level-25 green is about
# 160 units tall with its headline, and a PrintWindow capture of a live client
# measures the free band on the right - under the quest tracker, above the bag
# bar - at roughly 470 units. Three tooltips do not fit. Two do, with the third
# drop keeping its headline alone so nobody loses the attribution, which is
# half of what was asked for. A busy pull therefore costs one line of text
# rather than burying the fight.
MAX_CARDS = 3
MAX_EXPANDED = 2

# Group loot says the same drop twice - the roll result, then the loot - and
# the gap between them is a fraction of a second. Fifteen seconds is far wider
# than that gap and far narrower than the time it takes for the same character
# to genuinely loot a second copy of the same item.
DEDUPE_SECONDS = 15

# The headline's font size, and it is here rather than in the Lua because
# "readable on a phone watching the video" is a requirement rather than a
# palette choice. 18 with an outline against a 12-point tooltip underneath it.
# This is the number most likely to want tuning once somebody has actually
# watched it, which is the other reason it is written down somewhere tested.
HEADLINE_SIZE = 18

# Where the stack sits, as an offset from UIParent's BOTTOMRIGHT, growing UP.
#
# MEASURED, NOT GUESSED. On a live client the right column holds the buffs and
# the minimap down to y=200, the quest tracker to y=345, and then nothing at
# all until the action bar art begins around y=875. The bottom UI - bar, bags,
# micro menu - is about 103 units tall. 130 clears it with room, and growing
# upward means a busier stack extends into empty sky rather than down over the
# action bar. It is the opposite corner from the party frames and the chat
# log, so this and the status labels cannot collide.
#
# The one thing that would break it is turning on the right-hand vertical
# action bars, which are off on these clients and would occupy this column.
ANCHOR_X = -12
ANCHOR_Y = 130

# What a card says happened.
WON = "won"
LOOTED = "looted"

# (global name, which ARGUMENT is the winner, which is the item, what happened)
#
# Argument 0 means the message has no winner in it because it is one of the
# forms the client writes about itself. The addon substitutes the character it
# is watching, so a stream never says "You won" at a viewer who has no idea
# who "you" is.
MESSAGE_FORMS = (
    ("LOOT_ROLL_YOU_WON_NO_SPAM_NEED", 0, 2, WON),
    ("LOOT_ROLL_YOU_WON_NO_SPAM_GREED", 0, 2, WON),
    ("LOOT_ROLL_YOU_WON_NO_SPAM_DE", 0, 2, WON),
    ("LOOT_ROLL_WON_NO_SPAM_NEED", 1, 3, WON),
    ("LOOT_ROLL_WON_NO_SPAM_GREED", 1, 3, WON),
    ("LOOT_ROLL_WON_NO_SPAM_DE", 1, 3, WON),
    ("LOOT_ROLL_YOU_WON", 0, 1, WON),
    ("LOOT_ROLL_WON", 1, 2, WON),
    ("LOOT_ITEM_SELF_MULTIPLE", 0, 1, LOOTED),
    ("LOOT_ITEM_MULTIPLE", 1, 2, LOOTED),
    ("LOOT_ITEM_PUSHED_SELF_MULTIPLE", 0, 1, LOOTED),
    ("LOOT_ITEM_SELF", 0, 1, LOOTED),
    ("LOOT_ITEM", 1, 2, LOOTED),
    ("LOOT_ITEM_PUSHED_SELF", 0, 1, LOOTED),
)


def worth_showing(quality) -> bool:
    """Whether this drop earns a card.

    `quality` is what the client knows: 0 Poor through 6 Artifact, or None when
    the link carried no colour this build recognises.
    """
    if quality is None:
        return False
    return int(quality) >= MIN_QUALITY


def headline(winner: str, verb: str, viewer: str) -> str:
    """"Og won" - the words above the tooltip, without the item.

    The item is not in here because only the client can render it. The addon
    appends the link itself, which draws as a coloured `[Robe of the Moccasin]`
    for free; composing that name here would mean re-deriving the colour and
    the brackets and getting one of them wrong.

    NOBODY IS CALLED "YOU" ON A STREAM. The self forms carry no name, so the
    watched character's is used instead - a viewer looking at a video has no
    way to know whose screen it is otherwise, and "You won" beside four other
    people's names is the one sentence that could be about any of them.
    """
    who = (winner or "").strip() or (viewer or "").strip()
    if not who:
        return ""
    return "%s %s" % (who, verb)


def expire(stack: list[dict], now: float) -> list[dict]:
    """The cards still inside their dwell, oldest first."""
    return [c for c in stack if (now - float(c.get("at", 0))) < DWELL_SECONDS]


def admit(stack: list[dict], card: dict, now: float) -> list[dict]:
    """The stack after one drop arrives.

    INDEPENDENT TIMERS, and a new drop does NOT reset the old ones. Each card
    is a different moment and each gets its own couple of minutes; a version
    that restarted the clock on every drop would keep the first robe of a
    dungeon on screen until the run ended.

    The collapse of a duplicate keeps the ORIGINAL timestamp for the same
    reason. Group loot's second message about one drop must not buy that drop
    another two minutes.
    """
    kept = expire(list(stack), now)
    for existing in kept:
        if existing.get("link") != card.get("link"):
            continue
        if existing.get("winner") != card.get("winner"):
            continue
        if (now - float(existing.get("at", 0))) > DEDUPE_SECONDS:
            continue
        # Same drop, said twice. Upgrade the wording if the second telling is
        # the better one, and leave the clock alone.
        if existing.get("verb") == LOOTED and card.get("verb") == WON:
            existing["verb"] = WON
        return kept
    fresh = dict(card)
    fresh["at"] = now
    kept.append(fresh)
    # Oldest off the front: the newest drop is the one somebody is watching
    # for, so it must never be the one that is refused.
    return kept[-MAX_CARDS:]


def layout(stack: list[dict]) -> list[dict]:
    """Each card with whether it keeps its tooltip, oldest first.

    The NEWEST cards are the expanded ones. The addon draws the newest nearest
    the anchor and the rest above it, so the card a viewer is waiting for
    always appears in the same place instead of being pushed around by the ones
    before it.
    """
    out = []
    for i, card in enumerate(stack):
        entry = dict(card)
        entry["expanded"] = i >= len(stack) - MAX_EXPANDED
        out.append(entry)
    return out

File: wow-addons/PartyStatus/LootCard.lua
-- Loot Card (infra#3334, follow-up)
--
-- WHAT THIS IS. When the family wins something, a card carrying the item's own
-- tooltip and a large line saying who got it, for a couple of minutes. On a
-- stream the whole moment - what dropped, is it any good, who has it now - is
-- currently a truncated icon and a chat line that scrolls away, and nobody is
-- at the keyboard to hover.
--
-- THE STATS ARE THE GAME'S. GameTooltip:SetHyperlink on the link that came
-- through chat makes the client draw its own tooltip, so a suffix, an enchant
-- or a random roll is right because the client drew it. Nothing here composes
-- an item, and nothing here should: a display that builds its own stats starts
-- lying the first time an item is not the plain case.
--
-- WHERE THE JUDGEMENT LIVES, AND IT IS NOT HERE. The quality bar, the dwell,
-- how many cards may be up, which of them keep their tooltip, the words in the
-- headline, which chat lines are loot AND THE ORDER TO TRY THEM IN are all
-- decided in production/scripts/wow-overseer/lootcard.py and covered by its
-- stdlib suite. tests/test_lootcard.py reads this file as text and fails when
-- the two drift. This file turns those format strings into patterns and paints.
--
-- THE FORMAT STRINGS ARE THE CLIENT'S. LOOT_ITEM and friends only exist inside
-- a running client, which is why the pattern building cannot move to Python;
-- writing "%s receives loot: %s." in English here is exactly what those globals
-- exist to prevent.
--
-- WHERE IT SITS. Bottom right, growing UP, and measured against a PrintWindow
-- capture of a live client rather than guessed. That column holds the buffs and
-- the minimap down to y=200 and the quest tracker to y=345, then nothing at all
-- until the action bar art begins around y=875; the whole bottom UI - bar, bags,
-- micro menu - is about 103 units tall, which ANCHOR_Y clears. Growing upward
-- means a busy pull extends into empty sky rather than down over the action
-- bar, and it is the opposite corner from the party frames and the chat log, so
-- this and the status labels cannot reach each other. The one thing that would
-- break it is turning on the right-hand vertical action bars, which are off on
-- these clients and would occupy this column.

local lootAddon = {}

-- Mirrors lootcard.py. Every one of these is a decision made there.
local MIN_QUALITY = 2
local DWELL_SECONDS = 120
local MAX_CARDS = 3
local MAX_EXPANDED = 2
local DEDUPE_SECONDS = 15
local HEADLINE_SIZE = 18
local ANCHOR_X = -12
local ANCHOR_Y = 130
local WON = "won"
local LOOTED = "looted"

-- (global name, winner argument, item argument, what happened). Argument 0
-- means the message has no name in it because the client wrote it about
-- itself, and the watched character's name is used instead.
--
-- THE ORDER IS LOAD-BEARING. "%s won: %s" also matches the "(Need - 76)" line
-- with the suffix swallowed into the item capture, and matches "You won: ..."
-- with a winner called "You". Specific before general, self before third
-- person. Reorder it and the wrong capture wins silently.
local MESSAGE_FORMS = {
    { "LOOT_ROLL_YOU_WON_NO_SPAM_NEED", 0, 2, WON },
    { "LOOT_ROLL_YOU_WON_NO_SPAM_GREED", 0, 2, WON },
    { "LOOT_ROLL_YOU_WON_NO_SPAM_DE", 0, 2, WON },
    { "LOOT_ROLL_WON_NO_SPAM_NEED", 1, 3, WON },
    { "LOOT_ROLL_WON_NO_SPAM_GREED", 1, 3, WON },
    { "LOOT_ROLL_WON_NO_SPAM_DE", 1, 3, WON },
    { "LOOT_ROLL_YOU_WON", 0, 1, WON },
    { "LOOT_ROLL_WON", 1, 2, WON },
    { "LOOT_ITEM_SELF_MULTIPLE", 0, 1, LOOTED },
    { "LOOT_ITEM_MULTIPLE", 1, 2, LOOTED },
    { "LOOT_ITEM_PUSHED_SELF_MULTIPLE", 0, 1, LOOTED },
    { "LOOT_ITEM_SELF", 0, 1, LOOTED },
    { "LOOT_ITEM", 1, 2, LOOTED },
    { "LOOT_ITEM_PUSHED_SELF", 0, 1, LOOTED },
}

local CARD_GAP = 6
local CARD_WIDTH = 300

-- --- turning a format string into a pattern ----------------------------------

-- Lua's magic characters, MINUS % and $. % is left alone because the
-- specifiers are made of it, and a $ inside a pattern is literal anywhere but
-- the last character - which is the one this appends itself.
local MAGIC = "([%^%(%)%.%[%]%*%+%-%?])"

local function compile(fmt)
    -- Returns an anchored pattern plus, for each capture in text order, which
    -- ARGUMENT it is. Positional forms like "%1$s won: %3$s ... %2$d" put the
    -- arguments out of order on purpose, so the two are not the same list.
    local order = {}
    local body = string.gsub(fmt, MAGIC, "%%%1")
    body = string.gsub(body, "%%(%d?)%$?([sd])", function(idx, kind)
        local at = #order + 1
        order[at] = tonumber(idx) or at
        if kind == "d" then return "(%d+)" end
        return "(.-)"
    end)
    return "^" .. body .. "$", order
end

local function captureFor(order, argIndex)
    for i = 1, #order do
        if order[i] == argIndex then return i end
    end
    return nil
end

local forms = nil

local function buildForms()
    -- Built once, at login, because the globals do not exist before FrameXML
    -- has loaded. A form whose global is missing on this build is skipped
    -- rather than crashing the file: an absent string is one loot message that
    -- goes unread, and a hard error is every one of them.
    forms = {}
    for _, spec in ipairs(MESSAGE_FORMS) do
        local fmt = _G[spec[1]]
        if type(fmt) == "string" then
            local pattern, order = compile(fmt)
            local item = captureFor(order, spec[3])
            if item then
                table.insert(forms, {
                    pattern = pattern,
                    winner = spec[2] > 0 and captureFor(order, spec[2]) or nil,
                    item = item,
                    verb = spec[4],
                })
            end
        end
    end
end

-- --- reading one loot message ------------------------------------------------

local function qualityOf(link)
    -- The cache first, the link's own colour second. The colour is in every
    -- link the server sends, so it answers before the item is cached, which
    -- matters: the message arrives at the same moment the item does.
    local cached = select(3, GetItemInfo(link))
    if cached then return cached end
    local hex = string.match(link, "^|c(%x%x%x%x%x%x%x%x)")
    if not hex then return nil end
    hex = string.lower("|c" .. hex)
    for q = 0, 7 do
        local colour = ITEM_QUALITY_COLORS and ITEM_QUALITY_COLORS[q]
        if colour and colour.hex and string.lower(colour.hex) == hex then
            return q
        end
    end
    return nil
end

local function read(message)
    -- One chat line to { winner, link, verb }, or nil for anything that is not
    -- a loot line we care about. Money, rolls in progress and another addon's
    -- traffic all land here and all leave as nil.
    if not forms or not message then return nil end
    for _, form in ipairs(forms) do
        local caps = { string.match(message, form.pattern) }
        if caps[1] ~= nil then
            local field = caps[form.item]
            local link = field and string.match(field, "|c%x+|Hitem:.-|h.-|h|r")
            if link then
                return {
                    winner = form.winner and caps[form.winner] or "",
                    link = link,
                    verb = form.verb,
                }
            end
        end
    end
    return nil
end

-- --- the stack ---------------------------------------------------------------

local stack = {}
local enabled = true
local dirty = false

local function expire(now)
    local kept = {}
    for _, card in ipairs(stack) do
        if (now - card.at) < DWELL_SECONDS then table.insert(kept, card) end
    end
    local changed = #kept ~= #stack
    stack = kept
    return changed
end

local function admit(card, now)
    expire(now)
    for _, existing in ipairs(stack) do
        if existing.link == card.link and existing.winner == card.winner
                and (now - existing.at) <= DEDUPE_SECONDS then
            -- Group loot says one drop twice: the roll result, then the loot.
            -- Take the better wording and leave the clock alone, so the second
            -- telling cannot buy the same drop another two minutes.
            if existing.verb == LOOTED and card.verb == WON then
                existing.verb = WON
                existing.headline = card.headline
                dirty = true
            end
            return
        end
    end
    card.at = now
    table.insert(stack, card)
    while #stack > MAX_CARDS do
        -- Oldest off the front. The newest drop is the one somebody is
        -- watching for, so it must never be the one that is refused.
        table.remove(stack, 1)
    end
    dirty = true
end

-- --- painting ----------------------------------------------------------------

local anchor, cards = nil, {}

local function makeCard(i)
    local card = CreateFrame("Frame", nil, UIParent)
    card:SetWidth(CARD_WIDTH)
    card:SetHeight(HEADLINE_SIZE + CARD_GAP)
    card:EnableMouse(false)
    card:Hide()

    local head = card:CreateFontString(nil, "OVERLAY")
    -- OUTLINE, and large. This is read off a video on a phone, where a shadow
    -- is the first thing an encoder throws away and 12 point is unreadable.
    head:SetFont(STANDARD_TEXT_FONT or "Fonts\\FRIZQT__.TTF", HEADLINE_SIZE, "OUTLINE")
    head:SetPoint("TOPRIGHT", card, "TOPRIGHT", 0, 0)
    head:SetJustifyH("RIGHT")
    head:SetTextColor(1, 0.82, 0)
    card.head = head

    local tip = CreateFrame("GameTooltip", "PartyStatusLootTip" .. i, card,
                            "GameTooltipTemplate")
    -- MEDIUM, not the template's TOOLTIP. A card is furniture; a real tooltip
    -- somebody hovered must still draw over it.
    tip:SetFrameStrata("MEDIUM")
    tip:Hide()
    card.tip = tip
    return card
end

local function build()
    anchor = CreateFrame("Frame", nil, UIParent)
    anchor:SetWidth(1)
    anchor:SetHeight(1)
    anchor:SetPoint("BOTTOMRIGHT", UIParent, "BOTTOMRIGHT", ANCHOR_X, ANCHOR_Y)
    for i = 1, MAX_CARDS do cards[i] = makeCard(i) end
end

local function draw()
    if not anchor then return end
    local total = #stack
    for slot = 1, MAX_CARDS do
        local card = cards[slot]
        -- Slot 1 is the NEWEST and sits nearest the anchor, so the card a
        -- viewer is waiting for always appears in the same place instead of
        -- being shoved around by the ones before it.
        local entry = enabled and stack[total - slot + 1] or nil
        if not entry then
            card:Hide()
            card.tip:Hide()
        else
            card.head:SetText(entry.headline .. " " .. entry.link)
            local expanded = slot <= MAX_EXPANDED
            local height = HEADLINE_SIZE + CARD_GAP
            if expanded then
                card.tip:SetOwner(card, "ANCHOR_NONE")
                card.tip:ClearAllPoints()
                card.tip:SetPoint("TOPRIGHT", card, "TOPRIGHT", 0, -height)
                -- pcall: a malformed link throws, and one bad drop must cost
                -- its own card rather than every card after it.
                if pcall(card.tip.SetHyperlink, card.tip, entry.link) then
                    card.tip:Show()
                    height = height + card.tip:GetHeight()
                else
                    card.tip:Hide()
                end
            else
                card.tip:Hide()
            end
            card:SetHeight(height)
            card:ClearAllPoints()
            if slot == 1 then
                card:SetPoint("BOTTOMRIGHT", anchor, "BOTTOMRIGHT", 0, 0)
            else
                card:SetPoint("BOTTOMRIGHT", cards[slot - 1], "TOPRIGHT", 0, CARD_GAP)
            end
            card:Show()
        end
    end
end

-- --- driving -----------------------------------------------------------------

local TICK = 1.0
local since = 0
local settle = 0

local driver = CreateFrame("Frame")
driver:RegisterEvent("PLAYER_LOGIN")
driver:RegisterEvent("PLAYER_ENTERING_WORLD")
driver:RegisterEvent("CHAT_MSG_LOOT")

lootAddon.onLoot = function(message)
    if not enabled then return end
    local found = read(message)
    if not found then return end
    if not lootAddon.worthShowing(qualityOf(found.link)) then return end
    local who = found.winner
    if who == "" then who = UnitName("player") or "" end
    found.headline = who .. " " .. found.verb
    admit(found, GetTime())
end

lootAddon.worthShowing = function(quality)
    -- An item whose quality cannot be read is NOT shown. The colour is in
    -- every link the server sends, so an unreadable one means this reader is
    -- broken rather than that one unusual item arrived - and a broken reader
    -- that shows everything would put every grey on a permanent stream.
    if quality == nil then return false end
    return quality >= MIN_QUALITY
end

driver:SetScript("OnEvent", function(_, event, ...)
    if event == "CHAT_MSG_LOOT" then
        lootAddon.onLoot((...))
        return
    end
    if not forms then buildForms() end
    if not anchor then build() end
    draw()
end)

driver:SetScript("OnUpdate", function(_, elapsed)
    since = since + elapsed
    if dirty then
        dirty = false
        draw()
        -- Redraw once more next frame. A tooltip reports its height only after
        -- it has been laid out, so the first draw of a new card sizes the ones
        -- stacked above it against a height that is not final yet.
        settle = 2
    elseif settle > 0 then
        settle = settle - 1
        draw()
    end
    if since < TICK then return end
    since = 0
    if expire(GetTime()) then draw() end
end)

-- /loot        - hide the cards, or bring them back
-- /loot test   - put a card up for the first item in the bags, so the thing
--                can be seen without waiting for a dungeon to drop something
SLASH_PARTYSTATUSLOOT1 = "/loot"
SlashCmdList["PARTYSTATUSLOOT"] = function(arg)
    if arg == "test" then
        for bag = 0, 4 do
            for slot = 1, GetContainerNumSlots(bag) do
                local link = GetContainerItemLink(bag, slot)
                if link then
                    admit({ winner = UnitName("player") or "?", link = link,
                            verb = WON,
                            headline = (UnitName("player") or "?") .. " " .. WON },
                          GetTime())
                    DEFAULT_CHAT_FRAME:AddMessage("LootCard: showing " .. link)
                    return
                end
            end
        end
        DEFAULT_CHAT_FRAME:AddMessage("LootCard: nothing in the bags to show")
        return
    end
    enabled = not enabled
    dirty = true
    DEFAULT_CHAT_FRAME:AddMessage("LootCard: " .. (enabled and "on" or "off"))
end

-- Exposed on purpose. This is how a person at a client inspects what the
-- addon actually read out of a chat line, with /script and no debugger, and
-- it is how the local Lua harness drives all of it without a game running.
_G.PartyStatusLoot = lootAddon
lootAddon.read = read
lootAddon.compile = compile
lootAddon.stack = function() return stack end
lootAddon.cards = function() return cards end
